subsys/usb/usb_c/usbc_pe_src_states.c

```diff
@@ -0,0 +1,755 @@
+			}
+		}
+		/*
+		 * The Policy Engine Shall transition to the PE_SRC_Hard_Reset
+		 * state when:
+		 *	1) The SenderResponseTimer times out
+		 */
+		else if (usbc_timer_expired(&pe->pd_t_sender_response)) {
+			pe_set_state(dev, PE_SRC_HARD_RESET);
+		}
+		break;
+	}
+}
+
+void pe_src_send_capabilities_exit(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+
+	usbc_timer_start(&pe->pd_t_sender_response);
+}
+
+/**
+ * @brief 8.3.3.2.4 PE_SRC_Negotiate_Capability State
+ */
+void pe_src_negotiate_capability_entry(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+	struct usbc_port_data *data = dev->data;
+	struct protocol_layer_rx_t *prl_rx = data->prl_rx;
+
+	/* Get sink request */
+	pe->snk_request = *(uint32_t *)prl_rx->emsg.data;
+
+	/*
+	 * Ask the Device Policy Manager to evaluate the Request
+	 * from the Attached Sink.
+	 */
+	pe->snk_request_reply =
+		policy_check_sink_request(dev, pe->snk_request);
+
+	LOG_INF("PE_SRC_Negotiate_Capability");
+}
+
+void pe_src_negotiate_capability_run(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	/*
+	 * The Policy Engine Shall transition to the
+	 * PE_SRC_Transition_Supply state when:
+	 *	1) The Request can be met.
+	 */
+	if (pe->snk_request_reply == SNK_REQUEST_VALID) {
+		pe_set_state(dev, PE_SRC_TRANSITION_SUPPLY);
+	}
+	/*
+	 * The Policy Engine Shall transition to the
+	 * PE_SRC_Capability_Response state when:
+	 *	1) The Request cannot be met.
+	 *	2) Or the Request can be met later from the Power Reserve.
+	 */
+	else {
+		pe_set_state(dev, PE_SRC_CAPABILITY_RESPONSE);
+	}
+}
+
+/**
+ * @brief 8.3.3.2.5 PE_SRC_Transition_Supply State
+ */
+void pe_src_transition_supply_entry(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	LOG_INF("PE_SRC_Transition_Supply");
+
+	/*
+	 * If snk_request_reply is set, this state was entered
+	 * from PE_SRC_Negotiate_Capability. So send Accept Message
+	 * and inform the Device Policy Manager that it Shall transition
+	 * the power supply to the Requested power level.
+	 */
+	if (pe->snk_request_reply == SNK_REQUEST_VALID) {
+		send_ctrl_msg(dev, PD_PACKET_SOP, PD_CTRL_ACCEPT);
+	}
+	/*
+	 * If snk_request_reply is not valid, this state was entered
+	 * from PE_SRC_Ready. so send GotoMin Message.
+	 */
+	else {
+		send_ctrl_msg(dev, PD_PACKET_SOP, PD_CTRL_GOTO_MIN);
+	}
+}
+
+void pe_src_transition_supply_run(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	/*
+	 * The Policy Engine Shall transition to the PE_SRC_Ready state when:
+	 *	1) The Device Policy Manager informs the Policy Engine that
+	 *	   the power supply is ready.
+	 */
+	if (atomic_test_bit(&pe->flags, PE_FLAGS_TX_COMPLETE)) {
+		if (pe->snk_request_reply == SNK_REQUEST_VALID) {
```

Comment:
How do we transition out if we sent GotoMin?  (alternatively, we might leave out GotoMin for the first round on the source...)

Reply:
Good catch. I've updated the code.